src/main/java/org/spongepowered/mod/mixin/core/block/MixinBlockLeaves.java

```diff
@@ -53,26 +54,28 @@
     @Redirect(method = "breakBlock", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;beginLeavesDecay(Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;)V", remap = false))
     public void onBreakBlock(Block block, IBlockState state, net.minecraft.world.World worldIn, BlockPos pos) {
         if (!worldIn.isRemote) {
-            Sponge.getCauseStackManager().addContext(EventContextKeys.LEAVES_DECAY, (World) worldIn);
-            if (SpongeCommonEventFactory.callChangeBlockEventPre((IMixinWorldServer) worldIn, pos).isCancelled()) {
-                return;
-            }
-            final PhaseTracker phaseTracker = PhaseTracker.getInstance();
-            final IPhaseState currentState = phaseTracker.getCurrentState();
-            final boolean isBlockAlready = currentState.getPhase() != TrackingPhases.BLOCK;
-            final boolean isWorldGen = currentState.getPhase().isWorldGeneration(currentState);
-            if (isBlockAlready && !isWorldGen) {
-                final LocatableBlock locatable = LocatableBlock.builder()
-                        .location(new Location<World>((World) worldIn, pos.getX(), pos.getY(), pos.getZ()))
-                        .state((BlockState) state)
-                        .build();
-                BlockPhase.State.BLOCK_DECAY.createPhaseContext()
-                    .source(locatable)
-                    .buildAndSwitch();
-            }
-            block.beginLeavesDecay(state, worldIn, pos);
-            if (isBlockAlready && !isWorldGen) {
-                phaseTracker.completePhase(BlockPhase.State.BLOCK_DECAY);
+            try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+                Sponge.getCauseStackManager().addContext(EventContextKeys.LEAVES_DECAY, (World) worldIn);
+                if (SpongeCommonEventFactory.callChangeBlockEventPre((IMixinWorldServer) worldIn, pos).isCancelled()) {
+                    return;
+                }
+                final PhaseTracker phaseTracker = PhaseTracker.getInstance();
+                final IPhaseState currentState = phaseTracker.getCurrentState();
+                final boolean isBlockAlready = currentState.getPhase() != TrackingPhases.BLOCK;
+                final boolean isWorldGen = currentState.getPhase().isWorldGeneration(currentState);
+                if (isBlockAlready && !isWorldGen) {
+                    final LocatableBlock locatable = LocatableBlock.builder()
+                            .location(new Location<World>((World) worldIn, pos.getX(), pos.getY(), pos.getZ()))
+                            .state((BlockState) state)
+                            .build();
+                    BlockPhase.State.BLOCK_DECAY.createPhaseContext()
+                            .source(locatable)
+                            .buildAndSwitch();
+                }
+                block.beginLeavesDecay(state, worldIn, pos);
```

Comment:
This should be wrapped in a try with resources for the phase switch.

Reply:
Thanks for the advice I'm new to the phase stuff, but isn't the problem is the conditionals before and afterwards. 

All I've done is wrap everything in the if in a TryResource ContextFrame block.

I assume the !isWorldGen is to prevent spam during world generation as well as keeping the phase rules for world generation.

The isBlockAlready I don't quite understand.